Feature/circuit cmp

Skyfold opened the pull request:
This pull request needs to be done after the current feature/keccek pull request.

This just contains the code for comparison circuits. Currently I have written the `Word8` version (though its easy enough to write the others using it). I wanted to get some feedback on exactly how close to the Trello card I should be. For example, should I have a greater than zero function or just an example of how to combine a `const_word8` and the comparison function to check for greater than zero.

Commit:
nix: locked down the version of cargo and rustc

Commit:
inprogress, to show ross

ross-pure committed:
groth16/circuit.buillder: keccak helper functions

ross-pure committed:
Merge branch 'develop' of github.com:republicprotocol/zksnark-rs into develop

Commit:
builder: created Word64, KeccakMatrix types and keccack constants, lib: added itertools import

Commit:
field: Reverted to previous version

Commit:
circuit/builder: merged with develop

ross-pure committed:
Revert "Merge branch 'develop' of github.com:republicprotocol/zksnark-rs into develop"

This reverts commit aaa8edd58e835228da364503058f4147446607f4, reversing
changes made to 368089dfc0f54033c84754c4af4ffcd96927e6a6.

Commit:
circuit/builder: contains step0 keccak, needs testing, circuit/builder/word64: new file for Word64 logic and other keccak utils, Cargo.toml: added my name

Commit:
merged with develop: kept version at 0.0.1, otherwise no changes

Commit:
circuit/builder: switch to Iterators in progress, keccak step0 not refactored

Commit:
circuit/builder: still in progress, fixed rotate and added tests, step0 not done

Commit:
Revert "circuit/builder: still in progress, fixed rotate and added tests, step0 not done"

Type alias is a better approach

Commit:
Revert "circuit/builder: switch to Iterators in progress, keccak step0 not refactored"

This reverts commit 23c343b8f5aab20149d18dd5906f44f120e0fa93.

Commit:
circuit/builder: FromIterator tested for Word64 types, step0 not tested

Commit:
circuit/builder: Deref is only for smart pointers, last commit before I switch to type aliases

Commit:
circuit/builder: Word64 complete and tested, mod refactored for Word64, test added quickcheck stub

Commit:
circuit/builder: keccak-f is written!, no tests

Commit:
shell.nix: added vscode

Commit:
circuit/builder: keccak-f index fixed, basic/useless test written

Commit:
circuit/builder: sha3_256 written, not tested, need bit_streams

Commit:
circuit/builder: inprogress, tests failing, hash internals issues

Commit:
circuit/builder: word64 structs becames type, removed little endian, tests not passing

Commit:
field: fixed warning of extra parens

Commit:
circuit/builder: keccakf_1600 is equivalent to tiny_keccak keccakf

Commit:
circuit/builder: renamed module word64 to keccak_types

Commit:
groth16 and lib: removed qap from verify

Commit:
circuit/builder: completed keccak256

Commit:
circuit/builder: u64 and u8 fan_in fixed

Commit:
circuit and circuit/builder: exported Word8 and Word64 types and fixed keccak256 doc test

Commit:
circuit: exported to_word64 and to_word8 from circuit/builder

Commit:
circuit/builder: Added helper functions and keccak256_stream

Commit:
circuit/builder/types: adjusted to_word64/8 to panic on too little input and made to_word64 pub

Commit:
circuit/builder: Created BitChecker and other helper functions

Commit:
field: created generic to and from field bits

Commit:
circuit: exported BitChecker

ross-pure committed:
Update src/groth16/circuit/builder/types.rs

circuit/builder/types: remove dead code

Co-Authored-By: Skyfold <Skyfold@users.noreply.github.com>

Commit:
circuit and circuit/builder: moved setout into squeeze and reduced exports

Commit:
Merge branch 'feature/keccek' of github.com:republicprotocol/zksnark-rs into feature/keccek

ross-pure committed:
Update src/groth16/circuit/builder/types.rs

circuit/builder/types: typo XD

Co-Authored-By: Skyfold <Skyfold@users.noreply.github.com>

Commit:
Merge branch 'feature/keccek' of github.com:republicprotocol/zksnark-rs into feature/keccek

Commit:
circuit/builder: Added flatten to types and adjusted several type signatures

Commit:
Merge branch 'feature/wrapper' into feature/keccek

Commit:
circuit/builder: removed commented out code, reverted minor change in lib.rs

Commit:
src/lib: single keccak256 circuit example in progress

Commit:
Merge branch 'feature/wrapper' into feature/keccek

ross-pure committed:
Update src/groth16/circuit/builder/mod.rs

circuit/builder: alphabetical order of derive

Co-Authored-By: Skyfold <Skyfold@users.noreply.github.com>

ross-pure committed:
Update src/groth16/circuit/builder/mod.rs

circuit/builder: typo

Co-Authored-By: Skyfold <Skyfold@users.noreply.github.com>

ross-pure committed:
Update src/groth16/circuit/builder/mod.rs

circuit/builder: typo

Co-Authored-By: Skyfold <Skyfold@users.noreply.github.com>

Commit:
circuit/builder: Word8 comparison circuit

Commit:
Merge branch 'feature/keccek' into feature/circuit-cmp

Commit:
circuit/builder: Word64 comparison circuit and docs

Commit:
circuit/builder: new_equal with BinaryInput export issue

Commit:
circuit/builder: evaluate generic refactor, fixed doctests

Commit:
circuit/builder: added evaluate doctests and removed unnecessary traits

Commit:
circuit/builder: add all generic comparison functions and tests

ross-pure commented on src/groth16/circuit/builder/mod.rs:
```diff
@@ -755,9 +913,333 @@ where
+                    let eq = self.new_equality(*l, *r);
+                    self.new_and(eq, acc)
+                }
+                Left(l) => {
```
I think it's a little strange that the equality circuit will fill in zeros when the inputs are different lengths. It makes more sense to me to either panic or maybe return false, but I think this circuit should probably assume that the inputs are equal length and panic when they are not.

loongy replied:
I agree, panicking is the best approach here.

ross-pure replied:
Sure, there may be times where a user would be working with different bus widths and then want to do some sort of comparison on them, but for that case I think it would be better to have a specific `padded_equal`, probably built on top of a `pad` function that could either pad a single input to a desired width or given two inputs return both of them where the shorter one has been padded to be as long as the other. I think this makes sense because these primitives that we are writing now are at the low level of functionality (even though we plan to use just a comparison in our larger system), and then convenience functions should be built on top. At this level I think the analogue could be assembly; sure it's not usually convenient to use, but if someone is writing assembly then they usually have a good reason and want full control and are willing to sacrifice some convenience.

zbaako replied:
I prefer the proposal to force both left and right inputs to be the same type. Overall, I think it’s more likely that the programmer will erroneously compare different types that need to compare W8 and W64. If we encounter such a situation, then we can create a padding function for conversion, as Ross suggested.

loongy replied:
I like the proposal to enforce left and right inputs to be the same type. Implicit type conversion is typically removed in languages that want to encourage correctness, and when building a zkSNARK we definitely want that. 

I think, while it’s possible we would want to compare W8 and W64, it’s far more likely that we would do this erroneously (in all our proofs we need in the near future, all our data is a W64 except for hashes which aren’t used in comparisons with other non-hashes). To allow comparisons between different types we can add the pad function if we need it.

Skyfold replied:
You are right, I should no implicitly "convert" the types to compare them. I'll switch it back to:

```
    pub fn is_equal<'a, Z, P>(&mut self, left: Z, right: Z) -> WireId
    where
        Z: IntoIterator<Item = &'a WireId> + BinaryInput,
```

Skyfold replied:
That and all the other comparison functions

ross-pure commented on src/groth16/circuit/builder/mod.rs:
```diff
@@ -755,9 +913,333 @@ where
+                cmp.push_back(self.new_greater_than(*l, *r));
+                eq.push_back(self.new_equality(*l, *r));
+            }
+            Left(l) => {
```
Again I think that this function should panic if the inputs are not the same length

loongy replied:
I agree, panicking is the best approach here.

Skyfold replied:
I think its much better for the comparison functions to work for all combinations of number types since the type system constrains the inputs for us to ensure all possible invocations of the function are meaningful. (See previous comment)

loongy replied:
Whatever we do to solve the other one can be done here.

ross-pure reviewed (changes requested):
Looks pretty good, just some optimisations and some assumptions that I would like to see.

Commit:
Optimizes binary circuit building blocks

Skyfold commented on src/groth16/circuit/builder/mod.rs:
```diff
@@ -618,18 +757,38 @@ where
+    }
+
+    /// Requires that both the left and right inputs are either 0 or 1
+    pub fn new_nor(&mut self, lhs: WireId, rhs: WireId) -> WireId {
```
done

ross-pure replied:
You could have created the `ab` term using `new_and` ;) not really important though

Skyfold replied:
xD I'll fix that!

Skyfold commented on src/groth16/circuit/builder/mod.rs:
```diff
@@ -618,18 +757,38 @@ where
+    }
+
+    /// Requires that both the left and right inputs are either 0 or 1
+    pub fn new_xnor(&mut self, lhs: WireId, rhs: WireId) -> WireId {
```
done

Skyfold commented on src/groth16/circuit/builder/mod.rs:
```diff
@@ -755,9 +913,333 @@ where
+    }
+
+    /// Requires that both the left and right inputs are either 0 or 1
+    fn new_greater_than(&mut self, left: WireId, right: WireId) -> WireId {
```
done

Skyfold commented on src/groth16/circuit/builder/mod.rs:
```diff
@@ -755,9 +913,333 @@ where
 
-    // fn new_4_bit_compare(&mut self, lhs: [WireId; 4], rhs: [WireId; 4], cmp: [WireId; 3]) -> [WireId; 3] {
+    /// Requires that both the left and right inputs are either 0 or 1
+    fn new_less_than(&mut self, left: WireId, right: WireId) -> WireId {
```
done

Commit:
circuit/builder: Makes cmp require same input type

Note: I also have the preliminary parts of validate_balance
and have written validate_order circuit

Skyfold committed:
Merge branch 'master' into feature/circuit-cmp
